[WIP] Update Spigot to 1.13

Katrix opened the pull request:
Most stuff should be valid now, haven't done any tests yet. Still need to figure out what to do with some stuff.

Would be neat if I could get some info back on some of these.
* How to handle stencils?
* What in the world does BlockResetSurfaceBrush do?
* How to handle slab staircase for spiral staircase?'
* At the moment Ink no longer exists. Is there any realistic way to get it back?
* Many of the block lists would be better served as custom tags. AFAIK, Spigot provides no way to provide custom tags. Would it be okay to use nms to add these?

Katrix committed:
First work on 1.13

Deamon5550 commented on src/main/java/com/thevoxelbox/voxelsniper/Sniper.java:
```diff
@@ -284,12 +245,7 @@ public boolean snipe(Action action, Material itemInHand, Block clickedBlock, Blo
-                    MetricsManager.increaseBrushUsage(sniperTool.getCurrentBrush().getName());
-                }
-                return result;
+				return sniperTool.getCurrentBrush().perform(snipeAction, snipeData, targetBlock, lastBlock);
```
Don't replace spaces with tabs

Deamon5550 commented on src/main/java/com/thevoxelbox/voxelsniper/Sniper.java:
```diff
@@ -1,5 +1,18 @@
 package com.thevoxelbox.voxelsniper;
 
+import java.util.LinkedList;
```
Don't reformat the imports randomly.

Deamon5550 commented on src/main/java/com/thevoxelbox/voxelsniper/VoxelSniper.java:
```diff
@@ -1,17 +1,86 @@
 package com.thevoxelbox.voxelsniper;
 
-import com.google.common.base.Preconditions;
-import com.thevoxelbox.voxelsniper.brush.*;
```
Don't expand the star import.

Deamon5550 reviewed (changes requested):
The random changes of spaces to tabs and import reordering makes this quite hard to follow, please fix that.

The stencil brush should probably be updated to use schematics to be inline with the sponge version and generally more useful. There is a very old closed PR from me which did that, you could use it as a reference if you wanted to apply that change on top of this.

Ink would indeed be gone in the traditional way, but a much more powerful form could be made using the blockstates. Something like being able to specify `rotation=north` and apply that to all blocks which accept a rotation would replace the use that ink covered and be incredibly powerful.

I don't see any updates to the commands here, do the old commands still work with specifying the materials as numbers and data values? Also since you've removed some of the functions that the command call without updating them I'm guessing this doesn't actually compile?

Katrix commented:
Sorry for the whitespace changes. Didn't notice them before now. Will fix the import order too. Does the project have any official order? Looked around and didn't find any.

As for commands, I haven't done any changes there yet.

For ink, think that could be done, although don't know how fast it would be as it would mean a try-catch block for each performer call (could maybe reduce the overhead a bit with a cache) as Bukkit doesn't expose block traits in their raw form as far as I know.

Katrix committed:
Replace all tabs with spaces

Currently replaces stuff I'm fairly certain I haven't touched too. Will fix later

Katrix committed:
More work and fixes

Deamon5550 commented:
The import order is everything except java, and then java in a block afterwards.

I'd not worry about ink right now, that'll be a much more complicated change than warrants being rolled up into this change.

Katrix committed:
More fixes. Added back ink performers, fixed most commands, some other misc fixes

Katrix committed:
Fix hand deprecation warning

Katrix committed:
Start work on schematic stuff

gabizou commented:
Any further work going to be done on updating this? If time permits, I just might sit down and update this in a month or two (if I'm being realistic, might end up sitting down to update end of summer given everything I have higher on my priority list).

Katrix commented:
Ah, sorry. Been getting sidetracked by other projects instead of this (mostly Ore). Don't want to say anything concretely, but I might still work some more on this in the summer. AFAIK, most stuff except stencils should be done (although I haven't tested anything). I don't think I have any uncommited stuff, but I can check later and commit them if I have.

ghost commented:
> Any further work going to be done on updating this? If time permits, I just might sit down and update this in a month or two (if I'm being realistic, might end up sitting down to update end of summer given everything I have higher on my priority list).

This is not needed unless you really want to update the original repository. A member (@pitcer) from our team ([MCParkour Development Team](https://github.com/mcparkournet)) is currently working on updating VoxelSniper to the newest version of Minecraft for our server's purposes. The current version already works for Minecraft 1.13 and we did not see any severe errors while testing it. The reason we did not create a pull request was mainly due to we wanted to change the code to fit our requirements so code style was changed and so as most of the code in general but functionality and features remained intact.

Feel free to take a look at our fork: [voxel-sniper-flattened](https://github.com/mcparkournet/voxel-sniper-flattened)

gabizou commented:
> This is not needed unless you really want to update the original repository.

It's better to have the original project updated. It's been an idea of ours to replace stencils with schematics as they're far more supported (stencils are such a legacy format at this point anyways). I can't attest to my interest in letting the project totally die, it's probably best to have the core updated for the overall public to benefit. There's already a few forks of VS that are floating around and all are quasi bugged in various ways from mumblings around various communities, so getting it right as the source project, I'd feel it's necessary to getting it updated.

Katrix committed:
Stencil list no longer red

Katrix committed:
Don't need flameable materials tag

Katrix committed:
I have no idea what this thing does

Let's delete it for now and bring it back later if needed

Katrix committed:
Delete spiral staircase for now. Let's rewrite it once everything else runs

Katrix committed:
Add implementation defaults

Katrix committed:
Should be able to up this

Katrix committed:
Specify api-version

Katrix committed:
Fix erosion brush

Katrix committed:
Use method to get falling materials

Katrix commented:
Status update I guess. So, where is this at the moment? Almost done I think. It compiles, and runs fine on servers from the limited testing I've done. There's a few things left to be done. The tests are failing, and I don't know enough about maven to fix it. I also temporarily deleted the staircase brush, so that will have to be reimplemented. Finally I need to test that I implemented the schematics correctly.

Katrix committed:
Import cleanup

Katrix committed:
Missed a few

Katrix committed:
Misc bug fixes

Katrix committed:
1.16 NoteBlock fix

Katrix committed:
Updated for 1.18
